components/SerialServer/src/serial.c

```diff
@@ -507,10 +526,18 @@ void pre_init(void)
         getchar_clients = calloc(num_getchar_clients, sizeof(getchar_client_t));
         for (int i = 0; i < num_getchar_clients; i++) {
             unsigned int badge = getchar_enumerate_badge(i);
-            assert(badge <= num_getchar_clients);
+            if (badge > MAX_BADGE) {
+                ZF_LOGE("Client badge %d exceeds range 0->%d", badge, MAX_BADGE);
+                getchar_clients[i].idx = -1;
+                continue;
+            }
+            getchar_clients[i].idx = i;
             getchar_clients[i].client_id = badge;
             getchar_clients[i].buf = getchar_buf(badge);
             getchar_clients[i].last_head = -1;
+            if (i == 0) {
+                active_client = badge;
```

Comment:
There is no guarantee that `active_client` is set here, because the `if (badge > MAX_BADGE) {` may trigger for the first loop iteration already. So `active_client` might be undefined - or well, the global variable is set to `0`. Might be better so initialize this with `-1`, as client[0] could be invalid.

Reply:
Seem "-1" is used for multi-client already, so "-2" should be used then to indicate nothing is active.

Reply:
Good idea. I initialized `active_client` to -2, and then checked against `-2` instead of `i==0`.